src/main/java/org/usfirst/frc4048/common/swervev2/SwerveDrivetrain.java

```diff
@@ -0,0 +1,170 @@
+        PID steerPid = PID.of(Constants.STEER_PID_P,Constants.STEER_PID_I,Constants.STEER_PID_D);
+        Gain driveGain = Gain.of(Constants.DRIVE_PID_FF_V,Constants.DRIVE_PID_FF_S);
+        Gain steerGain = Gain.of(Constants.STEER_PID_FF_V,Constants.STEER_PID_FF_S);
+        frontLeft = new SparkMaxSwerveModule(encodedSwerveSparkMaxFL, drivePid,steerPid, driveGain,steerGain,constraints);
+        frontRight = new SparkMaxSwerveModule(encodedSwerveSparkMaxFR, drivePid,steerPid, driveGain,steerGain,constraints);
+        backLeft = new SparkMaxSwerveModule(encodedSwerveSparkMaxBL, drivePid,steerPid, driveGain,steerGain,constraints);
+        backRight = new SparkMaxSwerveModule(encodedSwerveSparkMaxBR, drivePid,steerPid, driveGain,steerGain,constraints);
+        frontLeftDrive.setInverted(true);
+        frontRightDrive.setInverted(false);
+        backRightDrive.setInverted(false);
+        backLeftDrive.setInverted(true);
+        navxGyro.setAngleAdjustment(gyroOffset);
+    }
+
+    @Override
+    public void periodic() {
+        navxGyroValue = getGyro();
+    }
+    public float getRoll() {
+        return navxGyro.getRoll();
+    }
+    
+    private static final double MODULE_MAX_ANGULAR_ACCELERATION =
+            2 * Math.PI; // radians per second squared
+    public void drive(double xSpeed, double ySpeed, double rot, boolean fieldRelative) {
+        SwerveModuleState[] swerveModuleStates = kinematics.toSwerveModuleStates(
+                fieldRelative
+                        ? ChassisSpeeds.fromFieldRelativeSpeeds(xSpeed, ySpeed, rot, new Rotation2d(Math.toRadians(navxGyroValue)))
+                        : new ChassisSpeeds(xSpeed, ySpeed, rot));
+        SwerveDriveKinematics.desaturateWheelSpeeds(swerveModuleStates, Constants.MAX_VELOCITY);
+        setModuleStates(swerveModuleStates);
+    }
+    public void driveRaw(double speed){
+        SwerveModuleState swerveModuleState = new SwerveModuleState(speed, new Rotation2d(0));
+        frontLeft.setDesiredState(swerveModuleState);
+        frontRight.setDesiredState(swerveModuleState);
+        backLeft.setDesiredState(swerveModuleState);
+        backRight.setDesiredState(swerveModuleState);
+    }
+
+    public void setModuleStates(SwerveModuleState[] desiredStates) {
+        frontLeft.setDesiredState(desiredStates[0]);
+        frontRight.setDesiredState(desiredStates[1]);
+        backLeft.setDesiredState(desiredStates[2]);
+        backRight.setDesiredState(desiredStates[3]);
+    }
+    
+    public void stopMotor() {
+        frontLeftTurn.set(0.0);
+        frontLeftDrive.set(0.0);
+        frontRightTurn.set(0.0);
+        frontRightDrive.set(0.0);
+        backLeftTurn.set(0.0);
+        backLeftDrive.set(0.0);
+        backRightTurn.set(0.0);
+        backRightDrive.set(0.0);
+    }
+
+    public void SetRelEncZero(){
+        frontLeft.getSwerveMotor().resetRelEnc();
```

Comment:
Why reset only for FrontLeft?

Reply:
I dont know why I didnt add the other motors (done)